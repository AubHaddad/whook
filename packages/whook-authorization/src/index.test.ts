import { wrapHandlerWithAuthorization } from '.';
import { handler } from 'knifecycle';
import YError from 'yerror';
import {
  BEARER as BEARER_MECHANISM,
  BASIC as BASIC_MECHANISM,
} from 'http-auth-utils';
import type { Parameters, Dependencies } from 'knifecycle';
import type { WhookOperation } from '@whook/whook';

describe('wrapHandlerWithAuthorization', () => {
  const noopMock = jest.fn(
    async (_d: Dependencies, _p: Parameters, _o: WhookOperation) => ({
      status: 200,
    }),
  );
  const log = jest.fn();
  const authentication = {
    check: jest.fn(),
  };
  const NOOP_OPERATION: WhookOperation = {
    path: '/path',
    method: 'get',
    operationId: 'noopHandler',
    summary: 'Does nothing.',
    tags: ['system'],
    parameters: [],
    responses: {
      200: {
        description: 'Sucessfully did nothing!',
      },
    },
  };
  const NOOP_AUTHENTICATED_OPERATION: WhookOperation = {
    ...NOOP_OPERATION,
    security: [
      {},
      {
        bearerAuth: ['user', 'admin'],
      },
    ],
  };
  const NOOP_RESTRICTED_OPERATION: WhookOperation = {
    ...NOOP_OPERATION,
    security: [
      {
        bearerAuth: ['user', 'admin'],
      },
    ],
  };
  const BAD_OPERATION: WhookOperation = {
    ...NOOP_OPERATION,
    security: [
      {
        bearerAuth: [],
      },
    ],
  };

  beforeEach(() => {
    noopMock.mockClear();
    log.mockReset();
    authentication.check.mockReset();
  });

  describe('with unauthenticated endpoints', () => {
    it('should work', async () => {
      const noopHandler = handler(noopMock, 'getNoop');
      const wrappedNoodHandlerWithAuthorization =
        wrapHandlerWithAuthorization(noopHandler);
      const wrappedHandler = await wrappedNoodHandlerWithAuthorization({
        authentication,
        log,
      });
      const response = await wrappedHandler({ aParameter: 1 }, NOOP_OPERATION);

      expect({
        response,
        noopMockCalls: noopMock.mock.calls,
        authenticationChecks: authentication.check.mock.calls,
        logCalls: log.mock.calls.filter(([type]) => !type.endsWith('stack')),
      }).toMatchSnapshot();
    });
  });

  describe('with authenticated but not restricted endpoints', () => {
    it('should work with bearer tokens and good authentication check', async () => {
      authentication.check.mockResolvedValue({
        applicationId: 'abbacaca-abba-caca-abba-cacaabbacaca',
        userId: 1,
        scope: 'user,admin',
      });
      const noopHandler = handler(noopMock, 'getNoop');
      const wrappedNoodHandlerWithAuthorization =
        wrapHandlerWithAuthorization(noopHandler);
      const wrappedHandler = await wrappedNoodHandlerWithAuthorization({
        authentication,
        log,
      });
      const response = await wrappedHandler(
        {
          authorization: 'bearer yolo',
        },
        NOOP_AUTHENTICATED_OPERATION,
      );

      expect({
        response,
        noopMockCalls: noopMock.mock.calls,
        authenticationChecks: authentication.check.mock.calls,
        logCalls: log.mock.calls.filter(([type]) => !type.endsWith('stack')),
      }).toMatchSnapshot();
    });

    it('should work with Bearer tokens and good authentication check', async () => {
      authentication.check.mockResolvedValue({
        applicationId: 'abbacaca-abba-caca-abba-cacaabbacaca',
        userId: 1,
        scope: 'user,admin',
      });
      const noopHandler = handler(noopMock, 'getNoop');
      const wrappedNoodHandlerWithAuthorization =
        wrapHandlerWithAuthorization(noopHandler);
      const wrappedHandler = await wrappedNoodHandlerWithAuthorization({
        authentication,
        log,
      });
      const response = await wrappedHandler(
        {
          authorization: 'Bearer yolo',
        },
        NOOP_AUTHENTICATED_OPERATION,
      );

      expect({
        response,
        noopMockCalls: noopMock.mock.calls,
        authenticationChecks: authentication.check.mock.calls,
        logCalls: log.mock.calls.filter(([type]) => !type.endsWith('stack')),
      }).toMatchSnapshot();
    });

    it('should work with access tokens and good authentication check', async () => {
      authentication.check.mockResolvedValue({
        applicationId: 'abbacaca-abba-caca-abba-cacaabbacaca',
        userId: 1,
        scope: 'user,admin',
      });
      const noopHandler = handler(noopMock, 'getNoop');
      const wrappedNoodHandlerWithAuthorization =
        wrapHandlerWithAuthorization(noopHandler);
      const wrappedHandler = await wrappedNoodHandlerWithAuthorization({
        authentication,
        log,
      });
      const response = await wrappedHandler(
        {
          access_token: 'yolo',
        },
        NOOP_AUTHENTICATED_OPERATION,
      );

      expect({
        response,
        noopMockCalls: noopMock.mock.calls,
        authenticationChecks: authentication.check.mock.calls,
        logCalls: log.mock.calls.filter(([type]) => !type.endsWith('stack')),
      }).toMatchSnapshot();
    });

    it('should work with no authentication at all', async () => {
      authentication.check.mockResolvedValue({
        applicationId: 'abbacaca-abba-caca-abba-cacaabbacaca',
        userId: 1,
        scope: 'user,admin',
      });
      const noopHandler = handler(noopMock, 'getNoop');
      const wrappedNoodHandlerWithAuthorization =
        wrapHandlerWithAuthorization(noopHandler);
      const wrappedHandler = await wrappedNoodHandlerWithAuthorization({
        authentication,
        log,
      });
      const response = await wrappedHandler({}, NOOP_AUTHENTICATED_OPERATION);

      expect({
        response,
        noopMockCalls: noopMock.mock.calls,
        authenticationChecks: authentication.check.mock.calls,
        logCalls: log.mock.calls.filter(([type]) => !type.endsWith('stack')),
      }).toMatchSnapshot();
    });
  });

  describe('with authenticated and restricted endpoints', () => {
    it('should work with bearer tokens and good authentication check', async () => {
      authentication.check.mockResolvedValue({
        applicationId: 'abbacaca-abba-caca-abba-cacaabbacaca',
        userId: 1,
        scope: 'user,admin',
      });
      const noopHandler = handler(noopMock, 'getNoop');
      const wrappedNoodHandlerWithAuthorization =
        wrapHandlerWithAuthorization(noopHandler);
      const wrappedHandler = await wrappedNoodHandlerWithAuthorization({
        authentication,
        log,
      });
      const response = await wrappedHandler(
        {
          authorization: 'Bearer yolo',
        },
        NOOP_RESTRICTED_OPERATION,
      );

      expect({
        response,
        noopMockCalls: noopMock.mock.calls,
        authenticationChecks: authentication.check.mock.calls,
        logCalls: log.mock.calls.filter(([type]) => !type.endsWith('stack')),
      }).toMatchSnapshot();
    });

    it('should work with access tokens and good authentication check', async () => {
      authentication.check.mockResolvedValue({
        applicationId: 'abbacaca-abba-caca-abba-cacaabbacaca',
        userId: 1,
        scope: 'user,admin',
      });
      const noopHandler = handler(noopMock, 'getNoop');
      const wrappedNoodHandlerWithAuthorization =
        wrapHandlerWithAuthorization(noopHandler);
      const wrappedHandler = await wrappedNoodHandlerWithAuthorization({
        authentication,
        log,
      });
      const response = await wrappedHandler(
        {
          access_token: 'yolo',
        },
        NOOP_RESTRICTED_OPERATION,
      );

      expect({
        response,
        noopMockCalls: noopMock.mock.calls,
        authenticationChecks: authentication.check.mock.calls,
        logCalls: log.mock.calls.filter(([type]) => !type.endsWith('stack')),
      }).toMatchSnapshot();
    });
  });

  it('should fail with no operation definition provided', async () => {
    const noopHandler = handler(noopMock, 'getNoop');
    const wrappedNoodHandlerWithAuthorization =
      wrapHandlerWithAuthorization(noopHandler);
    const wrappedHandler = await wrappedNoodHandlerWithAuthorization({
      authentication,
      log,
    });

    try {
      await wrappedHandler(
        {
          access_token: 'yolo',
        },
        undefined,
      );
      throw new YError('E_UNEXPECTED_SUCCESS');
    } catch (err) {
      expect({
        httpCode: err.httpCode,
        errorCode: err.code,
        errorParams: err.params,
        errorHeaders: err.headers,
        noopMockCalls: noopMock.mock.calls,
        authenticationChecks: authentication.check.mock.calls,
        logCalls: log.mock.calls.filter(([type]) => !type.endsWith('stack')),
      }).toMatchSnapshot();
    }
  });

  it('should fail with bad operation definition provided', async () => {
    const noopHandler = handler(noopMock, 'getNoop');
    const wrappedNoodHandlerWithAuthorization =
      wrapHandlerWithAuthorization(noopHandler);
    const wrappedHandler = await wrappedNoodHandlerWithAuthorization({
      authentication,
      log,
    });

    try {
      await wrappedHandler(
        {
          access_token: 'yolo',
        },
        BAD_OPERATION,
      );
      throw new YError('E_UNEXPECTED_SUCCESS');
    } catch (err) {
      expect({
        httpCode: err.httpCode,
        errorCode: err.code,
        errorParams: err.params,
        errorHeaders: err.headers,
        noopMockCalls: noopMock.mock.calls,
        authenticationChecks: authentication.check.mock.calls,
        logCalls: log.mock.calls.filter(([type]) => !type.endsWith('stack')),
      }).toMatchSnapshot();
    }
  });

  it('should fail without right scopes', async () => {
    authentication.check.mockResolvedValue({
      applicationId: 'abbacaca-abba-caca-abba-cacaabbacaca',
      userId: 1,
      scope: '',
    });
    const noopHandler = handler(noopMock, 'getNoop');
    const wrappedNoodHandlerWithAuthorization =
      wrapHandlerWithAuthorization(noopHandler);
    const wrappedHandler = await wrappedNoodHandlerWithAuthorization({
      authentication,
      log,
    });

    try {
      await wrappedHandler(
        {
          authorization: 'Bearer yolo',
        },
        NOOP_RESTRICTED_OPERATION,
      );
      throw new YError('E_UNEXPECTED_SUCCESS');
    } catch (err) {
      expect({
        httpCode: err.httpCode,
        errorCode: err.code,
        errorParams: err.params,
        errorHeaders: err.headers,
        noopMockCalls: noopMock.mock.calls,
        authenticationChecks: authentication.check.mock.calls,
        logCalls: log.mock.calls.filter(([type]) => !type.endsWith('stack')),
      }).toMatchSnapshot();
    }
  });

  it('should fail with unallowed mechanisms', async () => {
    authentication.check.mockResolvedValue({
      applicationId: 'abbacaca-abba-caca-abba-cacaabbacaca',
      userId: 1,
      scope: '',
    });
    const noopHandler = handler(noopMock, 'getNoop');
    const wrappedNoodHandlerWithAuthorization =
      wrapHandlerWithAuthorization(noopHandler);
    const wrappedHandler = await wrappedNoodHandlerWithAuthorization({
      MECHANISMS: [BASIC_MECHANISM, BEARER_MECHANISM],
      authentication,
      log,
    });

    try {
      await wrappedHandler(
        {
          authorization: 'Basic yolo',
        },
        NOOP_RESTRICTED_OPERATION,
      );
      throw new YError('E_UNEXPECTED_SUCCESS');
    } catch (err) {
      expect({
        httpCode: err.httpCode,
        errorCode: err.code,
        errorParams: err.params,
        errorHeaders: err.headers,
        noopMockCalls: noopMock.mock.calls,
        authenticationChecks: authentication.check.mock.calls,
        logCalls: log.mock.calls.filter(([type]) => !type.endsWith('stack')),
      }).toMatchSnapshot();
    }
  });

  it('should fail with not supported auth', async () => {
    authentication.check.mockRejectedValue(
      new YError('E_UNEXPECTED_TOKEN_CHECK'),
    );

    const noopHandler = handler(noopMock, 'getNoop');
    const wrappedNoodHandlerWithAuthorization =
      wrapHandlerWithAuthorization(noopHandler);
    const wrappedHandler = await wrappedNoodHandlerWithAuthorization({
      authentication,
      log,
    });

    try {
      await wrappedHandler(
        {
          authorization: 'Whatever yolo',
        },
        NOOP_RESTRICTED_OPERATION,
      );
      throw new YError('E_UNEXPECTED_SUCCESS');
    } catch (err) {
      expect({
        httpCode: err.httpCode,
        errorCode: err.code,
        errorParams: err.params,
        errorHeaders: err.headers,
        noopMockCalls: noopMock.mock.calls,
        authenticationChecks: authentication.check.mock.calls,
        logCalls: log.mock.calls.filter(([type]) => !type.endsWith('stack')),
      }).toMatchSnapshot();
    }
  });

  it('should fail with no authorization at all for secured endpoints', async () => {
    authentication.check.mockRejectedValue(
      new YError('E_UNEXPECTED_TOKEN_CHECK'),
    );

    const noopHandler = handler(noopMock, 'getNoop');
    const wrappedNoodHandlerWithAuthorization =
      wrapHandlerWithAuthorization(noopHandler);
    const wrappedHandler = await wrappedNoodHandlerWithAuthorization({
      authentication,
      log,
    });

    try {
      await wrappedHandler({}, NOOP_RESTRICTED_OPERATION);
      throw new YError('E_UNEXPECTED_SUCCESS');
    } catch (err) {
      expect({
        httpCode: err.httpCode,
        errorCode: err.code,
        errorParams: err.params,
        errorHeaders: err.headers,
        noopMockCalls: noopMock.mock.calls,
        authenticationChecks: authentication.check.mock.calls,
        logCalls: log.mock.calls.filter(([type]) => !type.endsWith('stack')),
      }).toMatchSnapshot();
    }
  });

  it('should fail with access_token disabled', async () => {
    authentication.check.mockRejectedValue(
      new YError('E_UNEXPECTED_TOKEN_CHECK'),
    );

    const noopHandler = handler(noopMock, 'getNoop');
    const wrappedNoodHandlerWithAuthorization =
      wrapHandlerWithAuthorization(noopHandler);
    const wrappedHandler = await wrappedNoodHandlerWithAuthorization({
      DEFAULT_MECHANISM: '',
      authentication,
      log,
    });

    try {
      await wrappedHandler(
        {
          access_token: 'yolo',
        },
        NOOP_RESTRICTED_OPERATION,
      );
      throw new YError('E_UNEXPECTED_SUCCESS');
    } catch (err) {
      expect({
        httpCode: err.httpCode,
        errorCode: err.code,
        errorParams: err.params,
        errorHeaders: err.headers,
        noopMockCalls: noopMock.mock.calls,
        authenticationChecks: authentication.check.mock.calls,
        logCalls: log.mock.calls.filter(([type]) => !type.endsWith('stack')),
      }).toMatchSnapshot();
    }
  });

  it('should proxy authentication errors', async () => {
    authentication.check.mockRejectedValue(
      new YError('E_UNEXPECTED_TOKEN_CHECK'),
    );

    authentication.check.mockRejectedValue(new YError('E_UNAUTHORIZED'));

    const noopHandler = handler(noopMock, 'getNoop');
    const wrappedNoodHandlerWithAuthorization =
      wrapHandlerWithAuthorization(noopHandler);
    const wrappedHandler = await wrappedNoodHandlerWithAuthorization({
      authentication,
      log,
    });

    try {
      await wrappedHandler(
        {
          authorization: 'Bearer yolo',
        },
        NOOP_RESTRICTED_OPERATION,
      );
      throw new YError('E_UNEXPECTED_SUCCESS');
    } catch (err) {
      expect({
        httpCode: err.httpCode,
        errorCode: err.code,
        errorParams: err.params,
        errorHeaders: err.headers,
        noopMockCalls: noopMock.mock.calls,
        authenticationChecks: authentication.check.mock.calls,
        logCalls: log.mock.calls.filter(([type]) => !type.endsWith('stack')),
      }).toMatchSnapshot();
    }
  });
});
