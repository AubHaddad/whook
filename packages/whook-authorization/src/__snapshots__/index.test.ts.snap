// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`wrapHandlerWithAuthorization should fail with a mismatch between user and authenticated one 1`] = `
Object {
  "authenticationChecks": Array [
    Array [
      "bearer",
      Object {
        "hash": "yolo",
      },
    ],
  ],
  "errorCode": "E_UNAUTHORIZED",
  "errorHeaders": Object {},
  "errorParams": Array [
    1,
    3,
  ],
  "httpCode": 403,
  "logCalls": Array [
    Array [
      "debug",
      "üîê - Initializing the authorization wrapper for \\"getNoop\\".",
    ],
  ],
  "noopMockCalls": Array [],
}
`;

exports[`wrapHandlerWithAuthorization should fail with access_token disabled 1`] = `
Object {
  "authenticationChecks": Array [],
  "errorCode": "E_UNAUTHORIZED",
  "errorHeaders": Object {
    "www-authenticate": "Bearer realm=\\"Auth\\"",
  },
  "errorParams": Array [],
  "httpCode": 401,
  "logCalls": Array [
    Array [
      "debug",
      "üîê - Initializing the authorization wrapper for \\"getNoop\\".",
    ],
    Array [
      "debug",
      "üîê - No authorization found, locking access!",
    ],
  ],
  "noopMockCalls": Array [],
}
`;

exports[`wrapHandlerWithAuthorization should fail with bad operation definition provided 1`] = `
Object {
  "authenticationChecks": Array [],
  "errorCode": "E_MISCONFIGURATION",
  "errorHeaders": Object {},
  "errorParams": Array [
    "Bearer",
    Array [],
    "noopHandler",
  ],
  "httpCode": 500,
  "logCalls": Array [
    Array [
      "debug",
      "üîê - Initializing the authorization wrapper for \\"getNoop\\".",
    ],
  ],
  "noopMockCalls": Array [],
}
`;

exports[`wrapHandlerWithAuthorization should fail with no authorization at all for secured endpoints 1`] = `
Object {
  "authenticationChecks": Array [],
  "errorCode": "E_UNAUTHORIZED",
  "errorHeaders": Object {
    "www-authenticate": "Bearer realm=\\"Auth\\"",
  },
  "errorParams": Array [],
  "httpCode": 401,
  "logCalls": Array [
    Array [
      "debug",
      "üîê - Initializing the authorization wrapper for \\"getNoop\\".",
    ],
    Array [
      "debug",
      "üîê - No authorization found, locking access!",
    ],
  ],
  "noopMockCalls": Array [],
}
`;

exports[`wrapHandlerWithAuthorization should fail with no operation definition provided 1`] = `
Object {
  "authenticationChecks": Array [],
  "errorCode": "E_OPERATION_REQUIRED",
  "errorHeaders": Object {},
  "errorParams": Array [],
  "httpCode": 500,
  "logCalls": Array [
    Array [
      "debug",
      "üîê - Initializing the authorization wrapper for \\"getNoop\\".",
    ],
  ],
  "noopMockCalls": Array [],
}
`;

exports[`wrapHandlerWithAuthorization should fail with not supported auth 1`] = `
Object {
  "authenticationChecks": Array [],
  "errorCode": "E_UNKNOWN_AUTH_MECHANISM",
  "errorHeaders": Object {},
  "errorParams": Array [
    "Whatever yolo",
  ],
  "httpCode": 400,
  "logCalls": Array [
    Array [
      "debug",
      "üîê - Initializing the authorization wrapper for \\"getNoop\\".",
    ],
  ],
  "noopMockCalls": Array [],
}
`;

exports[`wrapHandlerWithAuthorization should fail with unallowed mechanisms 1`] = `
Object {
  "authenticationChecks": Array [],
  "errorCode": "E_UNALLOWED_AUTH_MECHANISM",
  "errorHeaders": Object {},
  "errorParams": Array [
    "Basic yolo",
  ],
  "httpCode": 400,
  "logCalls": Array [
    Array [
      "debug",
      "üîê - Initializing the authorization wrapper for \\"getNoop\\".",
    ],
  ],
  "noopMockCalls": Array [],
}
`;

exports[`wrapHandlerWithAuthorization should fail without right scopes 1`] = `
Object {
  "authenticationChecks": Array [
    Array [
      "bearer",
      Object {
        "hash": "yolo",
      },
    ],
  ],
  "errorCode": "E_UNAUTHORIZED",
  "errorHeaders": Object {},
  "errorParams": Array [
    Array [],
    Array [
      "user",
      "admin",
    ],
  ],
  "httpCode": 403,
  "logCalls": Array [
    Array [
      "debug",
      "üîê - Initializing the authorization wrapper for \\"getNoop\\".",
    ],
  ],
  "noopMockCalls": Array [],
}
`;

exports[`wrapHandlerWithAuthorization should proxy authentication errors 1`] = `
Object {
  "authenticationChecks": Array [
    Array [
      "bearer",
      Object {
        "hash": "yolo",
      },
    ],
  ],
  "errorCode": "E_UNAUTHORIZED",
  "errorHeaders": Object {
    "www-authenticate": "Bearer realm=\\"Auth\\"",
  },
  "errorParams": Array [],
  "httpCode": 401,
  "logCalls": Array [
    Array [
      "debug",
      "üîê - Initializing the authorization wrapper for \\"getNoop\\".",
    ],
  ],
  "noopMockCalls": Array [],
}
`;

exports[`wrapHandlerWithAuthorization with authenticated and restricted endpoints should work with access tokens and good authentication check 1`] = `
Object {
  "authenticationChecks": Array [
    Array [
      "bearer",
      Object {
        "hash": "yolo",
      },
    ],
  ],
  "logCalls": Array [
    Array [
      "debug",
      "üîê - Initializing the authorization wrapper for \\"getNoop\\".",
    ],
  ],
  "noopMockCalls": Array [
    Array [
      Object {
        "authentication": Object {
          "check": [MockFunction] {
            "calls": Array [
              Array [
                "bearer",
                Object {
                  "hash": "yolo",
                },
              ],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": Promise {},
              },
            ],
          },
        },
        "log": [MockFunction] {
          "calls": Array [
            Array [
              "debug",
              "üîê - Initializing the authorization wrapper for \\"getNoop\\".",
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        },
      },
      Object {
        "access_token": "yolo",
        "authenticated": true,
        "scopes": Array [
          "user",
          "admin",
        ],
        "userId": 1,
      },
      Object {
        "method": "get",
        "operationId": "noopHandler",
        "path": "/path",
        "responses": Object {
          "200": Object {
            "description": "Sucessfully did nothing!",
          },
        },
        "security": Array [
          Object {
            "bearerAuth": Array [
              "user",
              "admin",
            ],
          },
        ],
        "summary": "Does nothing.",
        "tags": Array [
          "system",
        ],
      },
    ],
  ],
  "response": Object {
    "headers": Object {
      "X-Authenticated": "{\\"userId\\":1,\\"scopes\\":[\\"user\\",\\"admin\\"]}",
    },
    "status": 200,
  },
}
`;

exports[`wrapHandlerWithAuthorization with authenticated and restricted endpoints should work with bearer tokens and good authentication check 1`] = `
Object {
  "authenticationChecks": Array [
    Array [
      "bearer",
      Object {
        "hash": "yolo",
      },
    ],
  ],
  "logCalls": Array [
    Array [
      "debug",
      "üîê - Initializing the authorization wrapper for \\"getNoop\\".",
    ],
  ],
  "noopMockCalls": Array [
    Array [
      Object {
        "authentication": Object {
          "check": [MockFunction] {
            "calls": Array [
              Array [
                "bearer",
                Object {
                  "hash": "yolo",
                },
              ],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": Promise {},
              },
            ],
          },
        },
        "log": [MockFunction] {
          "calls": Array [
            Array [
              "debug",
              "üîê - Initializing the authorization wrapper for \\"getNoop\\".",
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        },
      },
      Object {
        "authenticated": true,
        "authorization": "Bearer yolo",
        "scopes": Array [
          "user",
          "admin",
        ],
        "userId": 1,
      },
      Object {
        "method": "get",
        "operationId": "noopHandler",
        "path": "/path",
        "responses": Object {
          "200": Object {
            "description": "Sucessfully did nothing!",
          },
        },
        "security": Array [
          Object {
            "bearerAuth": Array [
              "user",
              "admin",
            ],
          },
        ],
        "summary": "Does nothing.",
        "tags": Array [
          "system",
        ],
      },
    ],
  ],
  "response": Object {
    "headers": Object {
      "X-Authenticated": "{\\"userId\\":1,\\"scopes\\":[\\"user\\",\\"admin\\"]}",
    },
    "status": 200,
  },
}
`;

exports[`wrapHandlerWithAuthorization with authenticated but not restricted endpoints should work with Bearer tokens and good authentication check 1`] = `
Object {
  "authenticationChecks": Array [
    Array [
      "bearer",
      Object {
        "hash": "yolo",
      },
    ],
  ],
  "logCalls": Array [
    Array [
      "debug",
      "üîê - Initializing the authorization wrapper for \\"getNoop\\".",
    ],
  ],
  "noopMockCalls": Array [
    Array [
      Object {
        "authentication": Object {
          "check": [MockFunction] {
            "calls": Array [
              Array [
                "bearer",
                Object {
                  "hash": "yolo",
                },
              ],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": Promise {},
              },
            ],
          },
        },
        "log": [MockFunction] {
          "calls": Array [
            Array [
              "debug",
              "üîê - Initializing the authorization wrapper for \\"getNoop\\".",
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        },
      },
      Object {
        "authenticated": true,
        "authorization": "Bearer yolo",
        "scopes": Array [
          "user",
          "admin",
        ],
        "userId": 1,
      },
      Object {
        "method": "get",
        "operationId": "noopHandler",
        "path": "/path",
        "responses": Object {
          "200": Object {
            "description": "Sucessfully did nothing!",
          },
        },
        "security": Array [
          Object {},
          Object {
            "bearerAuth": Array [
              "user",
              "admin",
            ],
          },
        ],
        "summary": "Does nothing.",
        "tags": Array [
          "system",
        ],
      },
    ],
  ],
  "response": Object {
    "headers": Object {
      "X-Authenticated": "{\\"userId\\":1,\\"scopes\\":[\\"user\\",\\"admin\\"]}",
    },
    "status": 200,
  },
}
`;

exports[`wrapHandlerWithAuthorization with authenticated but not restricted endpoints should work with access tokens and good authentication check 1`] = `
Object {
  "authenticationChecks": Array [
    Array [
      "bearer",
      Object {
        "hash": "yolo",
      },
    ],
  ],
  "logCalls": Array [
    Array [
      "debug",
      "üîê - Initializing the authorization wrapper for \\"getNoop\\".",
    ],
  ],
  "noopMockCalls": Array [
    Array [
      Object {
        "authentication": Object {
          "check": [MockFunction] {
            "calls": Array [
              Array [
                "bearer",
                Object {
                  "hash": "yolo",
                },
              ],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": Promise {},
              },
            ],
          },
        },
        "log": [MockFunction] {
          "calls": Array [
            Array [
              "debug",
              "üîê - Initializing the authorization wrapper for \\"getNoop\\".",
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        },
      },
      Object {
        "access_token": "yolo",
        "authenticated": true,
        "scopes": Array [
          "user",
          "admin",
        ],
        "userId": 1,
      },
      Object {
        "method": "get",
        "operationId": "noopHandler",
        "path": "/path",
        "responses": Object {
          "200": Object {
            "description": "Sucessfully did nothing!",
          },
        },
        "security": Array [
          Object {},
          Object {
            "bearerAuth": Array [
              "user",
              "admin",
            ],
          },
        ],
        "summary": "Does nothing.",
        "tags": Array [
          "system",
        ],
      },
    ],
  ],
  "response": Object {
    "headers": Object {
      "X-Authenticated": "{\\"userId\\":1,\\"scopes\\":[\\"user\\",\\"admin\\"]}",
    },
    "status": 200,
  },
}
`;

exports[`wrapHandlerWithAuthorization with authenticated but not restricted endpoints should work with bearer tokens and good authentication check 1`] = `
Object {
  "authenticationChecks": Array [
    Array [
      "bearer",
      Object {
        "hash": "yolo",
      },
    ],
  ],
  "logCalls": Array [
    Array [
      "debug",
      "üîê - Initializing the authorization wrapper for \\"getNoop\\".",
    ],
  ],
  "noopMockCalls": Array [
    Array [
      Object {
        "authentication": Object {
          "check": [MockFunction] {
            "calls": Array [
              Array [
                "bearer",
                Object {
                  "hash": "yolo",
                },
              ],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": Promise {},
              },
            ],
          },
        },
        "log": [MockFunction] {
          "calls": Array [
            Array [
              "debug",
              "üîê - Initializing the authorization wrapper for \\"getNoop\\".",
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        },
      },
      Object {
        "authenticated": true,
        "authorization": "bearer yolo",
        "scopes": Array [
          "user",
          "admin",
        ],
        "userId": 1,
      },
      Object {
        "method": "get",
        "operationId": "noopHandler",
        "path": "/path",
        "responses": Object {
          "200": Object {
            "description": "Sucessfully did nothing!",
          },
        },
        "security": Array [
          Object {},
          Object {
            "bearerAuth": Array [
              "user",
              "admin",
            ],
          },
        ],
        "summary": "Does nothing.",
        "tags": Array [
          "system",
        ],
      },
    ],
  ],
  "response": Object {
    "headers": Object {
      "X-Authenticated": "{\\"userId\\":1,\\"scopes\\":[\\"user\\",\\"admin\\"]}",
    },
    "status": 200,
  },
}
`;

exports[`wrapHandlerWithAuthorization with authenticated but not restricted endpoints should work with no authentication at all 1`] = `
Object {
  "authenticationChecks": Array [],
  "logCalls": Array [
    Array [
      "debug",
      "üîê - Initializing the authorization wrapper for \\"getNoop\\".",
    ],
    Array [
      "debug",
      "üîì - Optionally authenticated enpoint detected, letting the call pass through!",
    ],
  ],
  "noopMockCalls": Array [
    Array [
      Object {
        "authentication": Object {
          "check": [MockFunction],
        },
        "log": [MockFunction] {
          "calls": Array [
            Array [
              "debug",
              "üîê - Initializing the authorization wrapper for \\"getNoop\\".",
            ],
            Array [
              "debug",
              "üîì - Optionally authenticated enpoint detected, letting the call pass through!",
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        },
      },
      Object {
        "authenticated": false,
      },
      Object {
        "method": "get",
        "operationId": "noopHandler",
        "path": "/path",
        "responses": Object {
          "200": Object {
            "description": "Sucessfully did nothing!",
          },
        },
        "security": Array [
          Object {},
          Object {
            "bearerAuth": Array [
              "user",
              "admin",
            ],
          },
        ],
        "summary": "Does nothing.",
        "tags": Array [
          "system",
        ],
      },
    ],
  ],
  "response": Object {
    "status": 200,
  },
}
`;

exports[`wrapHandlerWithAuthorization with unauthenticated endpoints should work 1`] = `
Object {
  "authenticationChecks": Array [],
  "logCalls": Array [
    Array [
      "debug",
      "üîê - Initializing the authorization wrapper for \\"getNoop\\".",
    ],
    Array [
      "debug",
      "üîì - Public endpoint detected, letting the call pass through!",
    ],
  ],
  "noopMockCalls": Array [
    Array [
      Object {
        "authentication": Object {
          "check": [MockFunction],
        },
        "log": [MockFunction] {
          "calls": Array [
            Array [
              "debug",
              "üîê - Initializing the authorization wrapper for \\"getNoop\\".",
            ],
            Array [
              "debug",
              "üîì - Public endpoint detected, letting the call pass through!",
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        },
      },
      Object {
        "aParameter": 1,
        "authenticated": false,
      },
      Object {
        "method": "get",
        "operationId": "noopHandler",
        "path": "/path",
        "responses": Object {
          "200": Object {
            "description": "Sucessfully did nothing!",
          },
        },
        "summary": "Does nothing.",
        "tags": Array [
          "system",
        ],
      },
    ],
  ],
  "response": Object {
    "status": 200,
  },
}
`;
